www/docs/server/context.md

```diff
@@ -60,11 +62,88 @@ export const router = t.router;
 
 /**
  * Unprotected procedure
- **/
+ */
 export const publicProcedure = t.procedure;
 
 /**
  * Protected procedure
- **/
+ */
 export const protectedProcedure = t.procedure.use(isAuthed);
-```
\ No newline at end of file
+```
+
+## Inner and outer context
+
+In some scenarios it could make sense to split up your context into "inner" and "outer" functions.
+
+**Inner context** is where you define context which doesn’t depend on the request, e.g. your database connection. You can use this function for integration testing or [SSG helpers](ssg-helpers), where you don’t have a request object. Whatever is defined here will **always** be available in your procedures.
+
+**Outer context** is where you define context which depends on the request, e.g. for the user's session. Whatever is defined here is not available in all procedures, but only for procedures that are called via HTTP.
+
+### Example for inner & outer context
+
+```ts
+import type { inferAsyncReturnType } from '@trpc/server';
+import type { CreateNextContextOptions } from '@trpc/server/adapters/next';
+
+/** 
+ * Defines your inner context shape.
+ * Add fields here that the inner context brings.
+ */
+interface CreateInnerContextOptions extends Partial<CreateNextContextOptions> {
+  prisma: typeof prisma
+}
+
+/** 
+ * Inner context. Will always be available in your procedures, in contrast to the outer context.
+ * 
+ * Also useful for:
+ * - testing, so you don't have to mock Next.js' `req`/`res`
+ * - tRPC's `createSSGHelpers` where we don't have `req`/`res`
+ * 
+ * @see https://trpc.io/docs/context#inner-and-outer-context
+ */
+export async function createContextInner(opts?: CreateInnerContextOptions) {
+  return {
+    ...opts,
+    prisma,
+  }
+};
+
+/** 
+ * Outer context. Will e.g. bring `req` & `res` to the context as "not `undefined`".
+ * 
+ * @see https://trpc.io/docs/context#inner-and-outer-context
+ */
+export async function createContext(opts: CreateNextContextOptions) {
+  const contextInner = await createContextInner();
+
+  return {
+    ...contextInner,
+    req: opts.req,
+    res: opts.res,
+  }
+};
+
+export type Context = inferAsyncReturnType<typeof createContextInner>;
+
+// The usage in your router is the same as the example above.
+```
+
+It is important to infer your `Context` from the **inner** context, as only what is defined there is really always available in your procedures.
+
+If you don't want to check `req` or `res` for `undefined` in your procedures all the time, you could build a small reusable procedure for that:
```

Comment:
Is this important enough to be in the docs? Req and res being used inside of a procedure is a pattern that I've not seen much in tRPC codebases. Might be wrong about this though.

Reply:
I use `req` & `res` a lot when I don't want to evaluate the session in the context on every request - maybe this takes some time or needs a roundtrip to DB and this adds up to the general request time. Having such procedure now eliminates the need of an `undefined` check for `req` or `res` in your procedures all the time. The idea comes from KATT here: https://github.com/trpc/trpc/commit/5c999058cc86be4fd3f73785617084c1ff45cb1f

..but I'm not sure how common this is as well. Let me know if I should throw it out.

Reply:
I think it's a good addition. Our docs don't _guide_ people enough. Feel free to tweak as follow-ups if you want @c-ehrlich  :)